src/k8s/pkg/k8sd/app/app.go

```diff
@@ -151,6 +155,17 @@ func New(cfg Config) (*App, error) {
 	} else {
 		log.L().Info("feature-controller disabled via config")
 	}
+
+	if !cfg.DisableCSRSigningController {
+		app.csrsigningController = csrsigning.New(csrsigning.Options{
+			Snap:           cfg.Snap,
+			WaitReady:      app.readyWg.Wait,
+			LeaderElection: true,
```

Comment:
More than a suggestion, this is a question. This is a Kubernetes controller running outside the cluster, right? How does the leader election happen here in the cluster? Does it interact with the other K8s nodes to elect a controller leader?

Reply:
Right. Essentially, this maps to a `lease` object. See the "name" and "holder" columns

```
root@t1:~# k8s kubectl get leases -A
NAMESPACE         NAME                                   HOLDER                                                                      AGE
kube-node-lease   t1                                     t1                                                                          26h
kube-node-lease   t2                                     t2                                                                          26h
kube-system       a27980c4.k8sd-csrsigning-controller    t1_0fdf508a-72bf-40de-90a1-d967b34f3086                                     26h
kube-system       apiserver-tfvf6sfbh5ge4ia3ni5rdnxjoq   apiserver-tfvf6sfbh5ge4ia3ni5rdnxjoq_5612db5b-ece7-41fd-aafd-75e6d5fa833c   26h
kube-system       cilium-operator-resource-lock          t2-znc8m94nv6                                                               26h
kube-system       kube-controller-manager                t1_cbe995ef-cceb-4f9b-af4c-6a6cb25991cd                                     26h
kube-system       kube-scheduler                         t1_75da548f-bcca-4142-83c1-e9b76a738bd7                                     26h
```

What we want to avoid here, is having a 3-node control plane where the controller is running 3 times. In that case, any certificate would be signed thrice (with the 2 first getting discarded). Instead, we create a lease, and only let the owner of the lease run the code of the controller itself. Other controllers will stay dormant and watch for changes to this lease. If anything happens to the leader, the lease will expire, and someone else will take it and run the controller.

This is also exactly what the Kubernetes components themselves do (see e.g. the kube-scheduler and kube-controller-manager leases)

Reply:
This is amazing. Thank you for the great explanation!